File: core/envelope-sim/src/test/java/fr/sncf/osrd/envelope_sim/EnvelopeMaintainSpeedTest.java
package fr.sncf.osrd.envelope_sim;

import static fr.sncf.osrd.envelope.EnvelopeShape.*;
import static fr.sncf.osrd.envelope_sim.SimpleContextBuilder.TIME_STEP;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.google.common.collect.ImmutableRangeMap;
import fr.sncf.osrd.envelope.Envelope;
import fr.sncf.osrd.envelope.EnvelopeShape;
import fr.sncf.osrd.envelope.part.EnvelopePart;
import fr.sncf.osrd.envelope_sim.pipelines.MaxEffortEnvelope;
import fr.sncf.osrd.envelope_sim.pipelines.MaxSpeedEnvelope;
import org.junit.jupiter.api.Test;
import java.util.List;

public class EnvelopeMaintainSpeedTest {
    @Test
    public void suddenSlope() {
        var stops = new double[] { };
        var envelopePath = new EnvelopePath(
                10000,
                new double[] { 0, 5000, 6000, 7000, 8000, 8500, 9000, 10000 },
                new double[] { 0, 40, -40, 0, 50, -50, 0 },
                ImmutableRangeMap.of());
        var testRollingStock = SimpleRollingStock.STANDARD_TRAIN;
        var context = new EnvelopeSimContext(testRollingStock, envelopePath, TIME_STEP,
                SimpleRollingStock.LINEAR_EFFORT_CURVE_MAP);

        var flatMRSP = Envelope.make(EnvelopePart.generateTimes(
                List.of(EnvelopeProfile.CONSTANT_SPEED),
                new double[] { 0, 10000 }, new double[] { 44.4, 44.4}
        ));
        var maxSpeedEnvelope = MaxSpeedEnvelope.from(context, stops, flatMRSP);
        var maxEffortEnvelope = MaxEffortEnvelope.from(context, 0, maxSpeedEnvelope);
        check(maxEffortEnvelope, new EnvelopeShape[][] {
                {INCREASING}, {CONSTANT},
                {DECREASING, INCREASING}, {CONSTANT},
                {DECREASING, INCREASING}, {CONSTANT}
        });
        assertTrue(maxEffortEnvelope.continuous);
    }
}

File: core/envelope-sim/src/test/kotlin/fr/sncf/osrd/envelope_sim/EnvelopeMaintainSpeedTest.kt
package fr.sncf.osrd.envelope_sim

import fr.sncf.osrd.envelope.Envelope.Companion.make
import fr.sncf.osrd.envelope.EnvelopeShape
import fr.sncf.osrd.envelope.part.EnvelopePart.Companion.generateTimes
import fr.sncf.osrd.envelope_sim.pipelines.SimStop
import fr.sncf.osrd.envelope_sim.pipelines.maxEffortEnvelopeFrom
import fr.sncf.osrd.envelope_sim.pipelines.maxSpeedEnvelopeFrom
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class EnvelopeMaintainSpeedTest {
    @Test
    fun suddenSlope() {
        val stops = listOf<SimStop>()
        val path =
            EnvelopeSimPathBuilder.buildNonElectrified(
                10000.0,
                doubleArrayOf(0.0, 5000.0, 6000.0, 7000.0, 8000.0, 8500.0, 9000.0, 10000.0),
                doubleArrayOf(0.0, 40.0, -40.0, 0.0, 50.0, -50.0, 0.0),
            )
        val testRollingStock = SimpleRollingStock.STANDARD_TRAIN
        val context =
            EnvelopeSimContext(
                testRollingStock,
                path,
                SimpleContextBuilder.TIME_STEP,
                SimpleRollingStock.LINEAR_EFFORT_CURVE_MAP,
            )

        val flatMRSP =
            make(
                generateTimes(
                    listOf(EnvelopeProfile.CONSTANT_SPEED),
                    doubleArrayOf(0.0, 10000.0),
                    doubleArrayOf(44.4, 44.4),
                )
            )
        val maxSpeedEnvelope = maxSpeedEnvelopeFrom(context, stops, flatMRSP)
        val maxEffortEnvelope = maxEffortEnvelopeFrom(context, 0.0, maxSpeedEnvelope)
        EnvelopeShape.check(
            maxEffortEnvelope,
            arrayOf(
                arrayOf(EnvelopeShape.INCREASING),
                arrayOf(EnvelopeShape.CONSTANT),
                arrayOf(EnvelopeShape.DECREASING, EnvelopeShape.INCREASING),
                arrayOf(EnvelopeShape.CONSTANT),
                arrayOf(EnvelopeShape.DECREASING, EnvelopeShape.INCREASING),
                arrayOf(EnvelopeShape.CONSTANT),
            ),
        )
        Assertions.assertTrue(maxEffortEnvelope.continuous)
    }
}
